lib/pry-doc/pry_ext/show_source_with_c_internals/symbol_extractor.rb

```diff
@@ -0,0 +1,106 @@
+class SymbolExtractor
+  class << self
+    attr_accessor :file_cache
+  end
+  @file_cache = {}
+
+  def extract_code(info)
+    if info.original_symbol.start_with?("#define")
+      extract_macro(info)
+    elsif info.original_symbol =~ /\s*(struct|enum)\s*/
+      extract_struct(info)
+    elsif info.original_symbol.start_with?("}")
+      extract_typedef_struct(info)
+    elsif info.original_symbol =~/^typedef.*;$/
+      extract_typedef_oneliner(info)
+    else
+      extract_function(info)
+    end
+  end
+
+  private
+
+  def extract_struct(info)
+    source_file = source_from_file(info.file)
+    offset = 1
+    loop do
+      code = source_file[info.line, offset].join
+      return code if balanced?(code)
+      offset += 1
+    end
+  end
+
+  def extract_typedef_struct(info)
+    source_file = source_from_file(info.file)
+    offset = 1
+    loop do
+      code = source_file[info.line - offset..info.line].join
+      return code if balanced?(code)
+      offset += 1
+    end
+  end
+
+  def extract_macro(info)
+    source_file = source_from_file(info.file)
+    offset = 1
+    loop do
+      code = source_file[info.line, offset].join
+      return code unless source_file[info.line + offset - 1].strip.end_with?('\\')
+      offset += 1
+    end
+  end
+
+  def extract_typedef_oneliner(info)
+    source_file = source_from_file(info.file)
+    return source_file[info.line]
+  end
+
+  def extract_function(info)
+    source_file = source_from_file(info.file)
+    offset = 1
+
+    if source_file[info.line] !~ /\w+\s+\w\(/ && source_file[info.line - 1].strip =~ /[\w\*]$/
```

Comment:
this won't work in the situation we have:

```c
char * foo() {
}
```

Need to relax the first regex so it looks like this:

``ruby
!~ /\w+\s*\*?\s+\w\(/ 
```

or some such